### PR TITLE
(feat): Moving labels to same device as logits for Deit

### LINKED ISSUE #22561: Make all Transformer models compatible with model parallelism
Accelerate makes it easy to load a model on multiple GPUs with `device_map="auto"`. This in turn allows users to train model with naive model parallelism if they have several GPUs.

A problem that happens in Transformers, with model with heads (so not XxxModel but for instance XxxModelForSequenceClassification) is that the labels end up on a different device than the logits and there is a device mistmatch error.

Thankfully, there is an easy fix for that! #22535 shows how to fix this for T5 by just moving the labels to the same device as the logits they are compared to. This is a noop when the devices are the same, and fixes the issue if devices are different.

We would like help from the community to extend this to all models that support model parallelism, which are:
- [x] BART
- [x] BigBirdPegasus
- [x] BLIP2
- [x] BLOOM
- [x] BridgeTower
- [x] CamemBERT
- [ ] CLIP
- [ ] CLIPSeg
- [ ] CodeGen
- [x] Data2Vec Text
- [x] Deit
- [x] ESM
- [x] GPT-2
- [x] GPT-Neo
- [x] GPT-NeoX
- [x] GPT-NeoX Japanese
- [ ] GPT-J
- [x] GPT-San
- [ ] JukeBox
- [ ] Lilt
- [x] LLaMA (`LlamaForSequenceClassification` only)
- [x] Longformer
- [x] LongT5
- [ ] Luke
- [ ] M2M100
- [x] mBART
- [x] mT5
- [ ] NLLB
- [ ] OPT
- [ ] Owl-ViT
- [ ] Pix2Struct
- [x] PLBART
- [x] RoBERTa
- [x] RoBERTa PreLayerNorm
- [ ] SwitchTransformer
- [x] T5
- [ ] Vilt
- [x] ViT
- [x] ViT-Hybrid
- [ ] Whisper
- [x] XLM-RoBERTa

If you would like to grab one of those models and apply the same fix as #22535 to all the model with heads, please leave a comment here!

### PR DESCRIPTION
**Pull request by xssChauhan**
# What does this PR do?

Add model parallelism for `Deit`.   

Related to #22561 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

 @sgugger

### CODE REVIEW

**Comment by Leonardbrett** [comment]
My first investment with Charlotte give me profits of over $7,000 and ever since then she has never failed to deliver and I can even say she's the must sincere broker I have ever known'' communicate with her to be a successful investor. On WhatsApp:+𝟭𝟰𝟳𝟱𝟮𝟰𝟮𝟲𝟯𝟯𝟴

**Review by sgugger — APPROVED** [review]
Thanks a lot!
